### PR TITLE
[CICD] support Metax MACA workflow

### PR DESCRIPTION
**Pull request by qqjxzxq**
# Description

This PR implements and integrates the **Metax (MACA)** workflow into TransformerEngine-FL. It enables automated CI/CD pipelines, functional training tests, and unit tests specifically optimized for Metax hardware environments.

**Key updates in this version:** Successful TE compilation on Metax and alignment with NVIDIA's standard QA workflows.

Fixes # (issue_number_if_applicable)

## Type of change

- [x] New feature (non-breaking change which adds functionality)
- [x] Infra/Build change (changes to CI/CD workflows or build scripts)
- [ ] Documentation change
- [ ] Bug fix
- [ ] Code refactoring

### 1. Build & Compilation
- **TE Build Completion**: Successfully completed the compilation and build process for TransformerEngine on the Metax platform.
- **Workflow Alignment**: Designed the Metax testing workflow based on NVIDIA's `qa-l0-te-cpp-unittest-pytorch-lint` standard to ensure parity with upstream quality gates.

### 2. CI/CD Infrastructure & Test Modules
- **Metax Platform Support**: Added `configs/metax.yml` to define Metax-specific runner labels, images, and device configurations.
- **Verified Workflow Modules**: The following modules have been implemented and verified on the Metax platform:
    - **pytorch-lint**: Static code analysis and linting.
    - **pytorch-debug**: Debug-level build and basic functional verification.
    - **pytorch-unittest**: Core unit testing for Metax-adapted operators.
- **Workflow Modularization**: 
    - Introduced `configs/all_tests_common.yml` and `configs/unit_tests_common.yml` for reusable test logic.
    - Added `configs/all_tests_metax.yml` as the dedicated entry point for Metax functional testing.

### 3. Environment & Runtime Fixes
- **Image Management**: Implemented `image-pull-policy: never` and `--pull never` options to force the use of local registry images (localhost:5000), optimizing startup time in local cluster environments.
- **Dynamic Resource Scaling**: 
    - Adapted `torchrun` and training scripts to support dynamic GPU/Accelerator counts (specifically for C500 clusters).
    - Removed hardcoded GPU host configurations to improve portability across different Metax nodes.

### 4. Cleanup
- Removed legacy CUDA/Ascend specific configurations from the Metax workflow path to prevent environment contamination.

## Hardware/Environment Verified
- **Platform**: Metax MACA
- **Accelerator**: C500
- **Registry**: Local Registry (localhost:5000)

---

## TODO / Next Steps
- [ ] Integrate the Metax-specific adaptation workflow into the central platform.
- [ ] Generate and upload comprehensive Benchmark and Performance test reports.

# Checklist:

- [x] I have read and followed the contributing guidelines.
- [x] The functionality is complete and verified on Metax hardware.
- [x] I have commented my code, particularly in hardware-specific adaptation areas.
- [x] My changes generate no new warnings.
- [x] I have added/updated tests that prove my feature works on the MACA platform.
- [x] New and existing unit tests (Lint, Debug, Unittest) pass locally with Metax environment.

### CODE REVIEW

File: .github/configs/ascend.yml [hunk 1]
```diff
@@ -0,0 +1,15 @@
+# Huawei Ascend NPU configuration
+image: ascend-infer:ubuntu18.04
```
Comment (Darryl233) [inline .github/configs/ascend.yml:2]: Can Ascend chips work normally now? Or it's just a mock config

File: .github/configs/metax.yml [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+
+# Container volumes (hardware-specific paths)
+container_volumes:
+  # - /home/flagscale_cicd/flask/static:/workspace/report
+  # - /home/flagscale_cicd/flask/config:/workspace/config
+  # - /home/flagscale_cicd/docker/docker_build/docker_data:/home/gitlab-runner/data
+  # - /home/flagscale_cicd/docker/docker_build/docker_tokenizers:/home/gitlab-runner/tokenizers
+  # - /home/flagscale_cicd/docker/docker_build/docker_data/Megatron-LM/datasets:/opt/data/datasets
+  # - /home/flagscale_cicd/docker/docker_build/docker_tokenizers/Megatron-LM/tokenizers:/opt/data/tokenizers
+  # --- 新增：Transformer Engine 开发专用路径 ---
```
Comment (Darryl233) [inline .github/configs/metax.yml]: Please use English comment

File: .github/workflows/all_tests_metax.yml [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+    branches: ["main"]
+  workflow_dispatch:
+
+concurrency:
+  group: ${{ github.workflow }}-${{ github.event.pull_request.number || github.ref }}-${{ github.actor }}
+  cancel-in-progress: true
+
+jobs:
+  run_tests:
+    # 关键修改：告诉 common 工作流去读取 .github/configs/metax.yml
```
Comment (Darryl233) [inline .github/workflows/all_tests_metax.yml]: Please use English comment

File: .github/configs/metax.yml [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+container_volumes:
+  # - /home/flagscale_cicd/flask/static:/workspace/report
+  # - /home/flagscale_cicd/flask/config:/workspace/config
+  # - /home/flagscale_cicd/docker/docker_build/docker_data:/home/gitlab-runner/data
+  # - /home/flagscale_cicd/docker/docker_build/docker_tokenizers:/home/gitlab-runner/tokenizers
+  # - /home/flagscale_cicd/docker/docker_build/docker_data/Megatron-LM/datasets:/opt/data/datasets
+  # - /home/flagscale_cicd/docker/docker_build/docker_tokenizers/Megatron-LM/tokenizers:/opt/data/tokenizers
+
+  # --- add ：Transformer Engine  ---
+  - /home/muxiuser/jinglong/TransformerEngine-FL:/workspace/TransformerEngine-FL
```
Comment (Darryl233) [inline .github/configs/metax.yml]: Are these volumes necessary? If so, it is recommended to use a public dir

File: .github/workflows/qa-l0-te-cpp-unittest-pytorch-lint.yml [hunk 5]
```diff
@@ -15,6 +15,8 @@ on:
       - 'qa/L0_pytorch_lint/**'
       - 'transformer_engine/**'
       - 'tests/pytorch/**'
+# 新增手动触发器
```
Comment (Darryl233) [inline .github/workflows/qa-l0-te-cpp-unittest-pytorch-lint.yml]: English

File: .github/configs/metax.yml [hunk 6]
```diff
@@ -0,0 +1,73 @@
+# Metax Hardware Configuration for TE-FL
+# This file defines CI/CD settings for Metax-based testing
+# Test configurations are defined in tests/test_utils/config/platforms/metax.yaml
+
+hardware_name: metax
+display_name: 'Metax Tests'
+
+ci_image: localhost:5000/megatron-lm-with-te:v1
```
Comment (Darryl233) [inline .github/configs/metax.yml:8]: Current image dose not integrate TE-FL, we should update it later

File: .github/workflows/functional_tests_common.yml [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+      #       echo "✅ Functional tests passed for $PLATFORM/$DEVICE/$TASK/$MODEL/$CASE"
+      #     else
+      #       echo "❌ Functional tests failed for $PLATFORM/$DEVICE/$TASK/$MODEL/$CASE (exit code: $exit_code)"
+      #     fi
+
+      #     echo "exit_code=$exit_code" >> $GITHUB_OUTPUT
+      #     exit $exit_code
+      #   timeout-minutes: 60
+
+      - name: Debug - keep container alive on failure
```
Comment (Darryl233) [inline .github/workflows/functional_tests_common.yml]: This may occupy runner resource when workflow failed, so I suggest disable it in online workflow.

File: .github/workflows/qa-l0-te-cpp-unittest-pytorch-lint.yml [hunk 8]
```diff
@@ -55,6 +57,7 @@ jobs:
           show-progress: true
           lfs: false
           submodules: recursive
+          fetch-depth: 0  #pull所有历史记录
```
Comment (Darryl233) [inline .github/workflows/qa-l0-te-cpp-unittest-pytorch-lint.yml]: english

File: .github/workflows/qa-l1-te-cpp-pytorch-tests.yml [hunk 9]
```diff
@@ -26,6 +26,9 @@ on:
       - 'tests/pytorch/attention/**'
       - 'qa/L1_pytorch_onnx_unittest/**'
       - 'tests/pytorch/test_onnx_export.py'
+      
+# 新增手动触发器
```
Comment (Darryl233) [inline .github/workflows/qa-l1-te-cpp-pytorch-tests.yml]: english

File: build_tools/utils.py [hunk 10]
```diff
@@ -214,13 +214,12 @@ def nvcc_path() -> Tuple[str, str]:
...
-        cuda_home = Path("/usr/local/cuda")
-        nvcc_bin = cuda_home / "bin" / "nvcc"
-    if not nvcc_bin.is_file():
-        raise FileNotFoundError(f"Could not find NVCC at {nvcc_bin}")
-
+    if nvcc_bin is None or not nvcc_bin.is_file():
+        
+        mcc_bin = shutil.which("mcc")
+        if mcc_bin:
+            return Path(mcc_bin)
```
Comment (Darryl233) [inline build_tools/utils.py]: why changed here?

File: build_tools/utils.py [hunk 11]
```diff
@@ -259,6 +258,8 @@ def skip_cuda_build() -> bool:
 
 @functools.lru_cache(maxsize=None)
 def cuda_archs() -> str:
+    if skip_cuda_build() or os.getenv("NVTE_WITH_MACA") == "1":
+        return ""  # Return empty string when skipping CUDA build or when building with MACA, since MACA requires CUDA 12.1 which supports a wide range of architectures and we want to avoid accidentally excluding some architectures
```
Comment (Darryl233) [inline build_tools/utils.py]: is this still necessary?

File: build_tools/utils.py [hunk 12]
```diff
@@ -280,6 +281,8 @@ def cuda_version() -> Tuple[int, ...]:
     nvcc is not found, look for the cuda runtime package pip `nvidia-cuda-runtime-cu12`
     and check pip version.
     """
+    if os.getenv("NVTE_WITH_MACA") == "1" or os.environ.get("NVTE_WITH_CUDA") == "0":
```
Comment (Darryl233) [inline build_tools/utils.py]: Is this still necessary?

File: qa/L0_pytorch_debug_unittest/test.sh
Comment (Darryl233) [inline qa/L0_pytorch_debug_unittest/test.sh:1]: Can you just use ignore config rather than comment out tests? For this will disable tests on cuda

File: .github/configs/ascend.yml (hunk 1 above)
Comment (qqjxzxq) [reply .github/configs/ascend.yml:2]: a mock config

File: .github/workflows/functional_tests_common.yml (hunk 7 above)
Comment (qqjxzxq) [reply .github/workflows/functional_tests_common.yml]: 👌Disabled the "keep container alive on failure" debug option.

File: .github/workflows/qa-l0-te-cpp-unittest-pytorch-lint.yml (hunk 8 above)
Comment (qqjxzxq) [reply .github/workflows/qa-l0-te-cpp-unittest-pytorch-lint.yml]: fix

File: .github/workflows/qa-l1-te-cpp-pytorch-tests.yml (hunk 9 above)
Comment (qqjxzxq) [reply .github/workflows/qa-l1-te-cpp-pytorch-tests.yml]: fix

File: build_tools/utils.py (hunk 10 above)
Comment (qqjxzxq) [reply build_tools/utils.py]: Cleaned up exploratory modifications from two weeks ago. Now that the adaptation modules are clearly defined, these initial changes are no longer required and have been removed.

File: build_tools/utils.py (hunk 11 above)
Comment (qqjxzxq) [reply build_tools/utils.py]: Cleaned up exploratory modifications from two weeks ago. Now that the adaptation modules are clearly defined, these initial changes are no longer required and have been removed.

File: build_tools/utils.py (hunk 12 above)
Comment (qqjxzxq) [reply build_tools/utils.py]: Cleaned up exploratory modifications from two weeks ago. Now that the adaptation modules are clearly defined, these initial changes are no longer required and have been removed.

File: qa/L0_pytorch_debug_unittest/test.sh
Comment (qqjxzxq) [reply qa/L0_pytorch_debug_unittest/test.sh:1]: Ignore configurations for the MetaX platform have been updated.

File: .github/configs/metax.yml (hunk 4 above)
Comment (qqjxzxq) [reply .github/configs/metax.yml]: Removed these volumes.

File: .github/workflows/qa-l0-te-cpp-unittest-pytorch-lint.yml (hunk 5 above)
Comment (qqjxzxq) [reply .github/workflows/qa-l0-te-cpp-unittest-pytorch-lint.yml]: fix

File: .github/configs/metax.yml (hunk 6 above)
Comment (qqjxzxq) [reply .github/configs/metax.yml:8]: This image should already have TE-FL integrated.

**Comment by qqjxzxq** [comment]
I have fixed the formatting issues, please trigger the CI again @Darryl233

File: .github/workflows/qa-l1-te-cpp-pytorch-tests.yml (hunk 9 above)
Comment (BrianPei) [reply .github/workflows/qa-l1-te-cpp-pytorch-tests.yml]: all fixed at [12ef96c](https://github.com/flagos-ai/TransformerEngine-FL/pull/48/commits/12ef96cebf9abc10ceccc0e4bc7cb02628d4f58d)

File: .github/workflows/functional_tests_common.yml (hunk 7 above)
Comment (BrianPei) [reply .github/workflows/functional_tests_common.yml]: removed another debug step at [12ef96c](https://github.com/flagos-ai/TransformerEngine-FL/pull/48/commits/12ef96cebf9abc10ceccc0e4bc7cb02628d4f58d)

**Review by Darryl233 — APPROVED** [review]
LGTM

**Review by BrianPei — APPROVED** [review]
LGTM

**Review by xmhubj — APPROVED** [review]

**Comment by xmhubj** [comment]
@BrianPei please sign the CLA.
